[feature] grants for additional resources

a129722 opened the pull request:
This PR adds functionality for grants of type:

- ADD SEARCH OPTIMIZATION on schema
- future grants on stages
- external tables
- materialized views
- streams
- procedures
- functions
- file formats
- sequences

## Test Plan
* [x] acceptance tests for future grants, normal grants require resource implementations
* [x] unit tests for all new grant types

ryanking commented on pkg/resources/external_table_grant.go:
```diff
@@ -0,0 +1,213 @@
+	futureExternalTables := data.Get("on_future").(bool)
+	grantOption := data.Get("with_grant_option").(bool)
+
+	if (schemaName == "") && !futureExternalTables {
+		return errors.New("schema_name must be set unless on_future is true.")
+	}
+
+	if (externalTableName == "") && !futureExternalTables {
+		return errors.New("external_table_name must be set unless on_future is true.")
+	}
+	if (externalTableName != "") && futureExternalTables {
+		return errors.New("external_table_name must be empty if on_future is true.")
+	}
+
```
I think we would have a better experience if we pull these up to the schema level with [ValidateDiagFunc](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Schema.ValidateDiagFunc)

ryanking commented on pkg/resources/external_table_grant_acceptance_test.go:
```diff
@@ -0,0 +1,74 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestAccExternalTableGrantFuture(t *testing.T) {
```
```suggestion
func TestAccExternalTableGrant_future(t *testing.T) {
```

a129722 replied:
So I was looking at other examples of format for test names and the suggestion above conflicts with others.  I ended up with the name `TestAcc_ExternalTableFutureGrant` and applied that elsewhere.  Let me know if you'd like something different.

ryanking commented on pkg/resources/file_format_acceptance_test.go:
```diff
@@ -0,0 +1,74 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestAccFileFormatGrantFuture(t *testing.T) {
```
```suggestion
func TestAccFileFormatGrant_future(t *testing.T) {
```

a129722 replied:
See above

ryanking commented on pkg/resources/file_format_acceptance_test.go:
```diff
@@ -0,0 +1,74 @@
+)
+
+func TestAccFileFormatGrantFuture(t *testing.T) {
+	databaseName := strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
+	schemaName := strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
+	roleName := strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
```
```suggestion
	databaseName := acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
	schemaName := acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
	roleName := acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
```

a129722 replied:
See above

ryanking commented on pkg/resources/file_format_grant.go:
```diff
@@ -0,0 +1,206 @@
+	"file_format_name": {
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "The name of the file format on which to grant privileges immediately (only valid if on_future is unset).",
```
```suggestion
		Description: "The name of the file format on which to grant privileges immediately (only valid if on_future is false).",
```

a129722 replied:
Updated

ryanking commented on pkg/resources/file_format_grant.go:
```diff
@@ -0,0 +1,206 @@
+	futureFileFormats := data.Get("on_future").(bool)
+	grantOption := data.Get("with_grant_option").(bool)
+
+	if (schemaName == "") && !futureFileFormats {
+		return errors.New("schema_name must be set unless on_future is true.")
+	}
+
+	if (fileFormatName == "") && !futureFileFormats {
+		return errors.New("file_format_name must be set unless on_future is true.")
+	}
+	if (fileFormatName != "") && futureFileFormats {
+		return errors.New("file_format_name must be empty if on_future is true.")
+	}
+
```
As with above, it would be great to pull these up to the schema level.

a129722 replied:
Done

ryanking reviewed (changes requested):
Thanks for this, majorly helpful work.

I only read through the first dozen or so files, but I think most of my comments there apply to the rest. Take a look and click the re-request review when ready and I will review the rest.

a129722 commented on pkg/resources/database_acceptance_test.go:
```diff
@@ -14,8 +15,8 @@ func TestAccDatabase(t *testing.T) {
 
-	prefix := acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
-	prefix2 := acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
+	prefix := strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
```
This change is to make the acceptance tests compatible with accounts that have [QUOTED_IDENTIFIERS_IGNORE_CASE](https://docs.snowflake.com/en/sql-reference/parameters.html#quoted-identifiers-ignore-case) set to `true`.  This will cause the any identifiers used to be represented in all upper-case within any system catalog-based queries.

a129722 replied:
That is correct

ryanking replied:
Ok, if this is blocking we can do it for now and I will think of more sustainable ways to do this.

a129722 commented on pkg/resources/external_table_grant.go:
```diff
@@ -0,0 +1,213 @@
+	"external_table_name": {
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "The name of the external table on which to grant privileges immediately (only valid if on_future is unset).",
```
Updated

a129722 commented on pkg/resources/external_table_grant.go:
```diff
@@ -0,0 +1,213 @@
+		Description:   "When this is set to true and a schema_name is provided, apply this grant on all future external tables in the given schema. When this is true and no schema_name is provided apply this grant on all future external tables in the given database. The external_table_name and shares fields must be unset in order to use on_future.",
+		Default:       false,
+		ForceNew:      true,
+		ConflictsWith: []string{"external_table_name", "shares"},
```
To answer the two ways to interpret the question... 

1. Yes, it only really conflicting when `external_table_name` is set to true, but there's no good way aside from this to enforce within the schema itself.  
2. Future grants are not supported with shares

a129722 commented on pkg/resources/external_table_grant.go:
```diff
@@ -0,0 +1,213 @@
+	)
+	if _, ok := data.GetOk("external_table_name"); ok {
+		externalTableName = data.Get("external_table_name").(string)
+	}
```
Updated

a129722 commented on pkg/resources/external_table_grant.go:
```diff
@@ -0,0 +1,213 @@
+	}
+	if _, ok := data.GetOk("schema_name"); ok {
+		schemaName = data.Get("schema_name").(string)
+	}
```
Updated

a129722 commented:
@ryanking - this is ready for re-review.  i've also integrated changes from `main` and added examples for all resources.

ryanking reviewed (changes requested):
This is another partial review. I am going to hold off on reviewing more until we land https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/333, which will requires some rebasing here.

a129722 commented on pkg/resources/external_table_grant.go:
```diff
@@ -0,0 +1,198 @@
+		Optional:      true,
+		Description:   "The name of the external table on which to grant privileges immediately (only valid if on_future is false).",
+		ForceNew:      true,
+		ConflictsWith: []string{"on_future"},
```
A validation function only operates on the value itself, so I think the validation would need to go within the Create function like we had before.

ryanking replied:
Yeah you are right. I will complain to the team at hashicorp about the lack of a good validation hook for this.

ryanking reviewed (changes requested):
Can you merge the main branch and resolve conflicts?

Also, note that the refactor in https://github.com/chanzuckerberg/terraform-provider-snowflake/commit/ffe977ad2bcd3165b1821ae79f44dec7246aeb7f is going to require some changes to this PR to match.

a129722 commented:
@ryanking - main was merged again and I updated the code to remove the `ConflictsWith` schema declarations and added in the checks within the `Create` methods.

a129722 commented:
@ryanking  - I was wondering if we could push to get this reviewed/integrated soon.  There's been a good amount change on the main branch and re-integrating everything has been time consuming on my end.

ryanking commented:
@a129722 yeah, as soon as you can resolve the merge conflicts we will take a look

a129722 commented:
@ryanking  - i merged in main.  not sure why the check-mod step is failing.  When I build/test locally, there are no changes that show up.

ryanking commented:
Yeah there is something weird going on with `check-mod`. If you run `go get -u golang.org/x/tools` and `go mod tidy` it should fix it (not sure why we recently had to start doing this).

a129722 commented:
thanks for the tip - looks like things are passing now

ryanking commented on pkg/resources/file_format_acceptance_test.go:
```diff
@@ -0,0 +1,73 @@
+package resources_test
```
Should be name `file_format_grant_acceptance_test.go`

ryanking commented on pkg/resources/file_format_acceptance_test.go:
```diff
@@ -0,0 +1,73 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestAcc_FileFormatFutureGrant(t *testing.T) {
```
```suggestion
func TestAcc_FileFormatGrantFutureGrant(t *testing.T) {
```

ryanking reviewed (changes requested):
Very minor changes and then this is good to merge.

a129722 commented:
Done!

ryanking commented:
/ok-to-test sha=e6e35370d4c06a68ebe88526916765b1c55a2fa6

a129722 commented:
Thanks, @ryanking !
